presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java

```diff
@@ -254,24 +245,32 @@ private void checkAllFuturesDone()
     }
 
     @Override
-    public CompletableFuture<LookupSource> readSpilledLookupSource(Session session, int partition)
+    public synchronized CompletableFuture<LookupSource> readSpilledLookupSource(Session session, int partition)
```

Comment:
besides synchronization on loading next partitions, you also need a code to clean up and close no longer used partitions.